### ISSUE TITLE
Reuse existing cache when rebuilding cache

### ISSUE THREAD

**Issue by danielcompton** [borkdude/clj-kondo#705]
**Is your feature request related to a problem? Please describe.**

In CI and when running a full local lint, we build the whole cache before linting our project

```bash
clj-kondo --lint "$(clojure -Spath)" --cache > /dev/null
git ls-files | grep -E '.clj[cs]?$' | xargs clj-kondo --lint
```

Rebuilding the whole cache is a relatively heavyweight operation, but gives us the most complete linting results (I think?). It would be nice if we could reuse this cache between CI builds to save build time.

**Describe the solution you'd like**

I'm not sure if this makes sense, but if each cache entry for an immutable file had an identifier in the name or file, that would let clj-kondo know when it could skip rebuilding the cache.

The tricky part then becomes knowing which files can be considered immutable, and which ones can't. My suggestion would be to only cache JARs with a non SNAPSHOT version.

**Describe alternatives you've considered**

* Not building the cache in CI. I think this might mean we miss some lint warnings?
* Rebuilding the cache in CI every time. This is just slow, but nothing bad happens.

**Comment by borkdude** [comment]
If you're using clj-kondo like this on CI you don't really need a cache.

Note that you can merge these two operations:

```
clj-kondo --lint "$(clojure -Spath)" --cache > /dev/null
git ls-files | grep -E '.clj[cs]?$' | xargs clj-kondo --lint
```

into one:

```
clj-kondo --lint "$(clojure -Spath)" --config '{:output {:include-files ["^src"]}}'
```

The config option takes Java regexes which will be matches using `(re-find (re-pattern %) file-name)`.

The `--cache` option defaults to `true` now, so you don't need it anymore, unless you want to disable caching.

This will likely already save you time.

If that's not enough, you can consider linting your libraries separately and save them to a separate cache directory with `--cache-dir` and use a checksum. You can then check if that directory already exists using the checksum and copy it to .clj-kondo/.cache before you lint your own sources.

**Comment by borkdude** [comment]
So to recap, for CircleCI I would do something like this:

``` yaml
      - restore_cache:
          keys:
            - deps-{{ checksum "deps.edn" }}

      - run:
          name: Lint
          command: |
            if [ -d ~/clj-kondo-cache ] then; # contains deps restored from cache
              clj-kondo --lint src
           else
             clj-kondo --lint $(clojure -Spath)
           fi
      - save_cache:
          paths:
            - ~/.m2
            - ~/clj-kondo-cache
          key: v1-dependencies-{{ checksum "deps.edn" }}
```

Note that I've not tested this, take it as pseudo-code.

**Comment by borkdude** [comment]
@danielcompton If the above works out for you, it might be nice to add the final version to this as documentation to: https://github.com/borkdude/clj-kondo/blob/master/doc/ci-integration.md

**Comment by danielcompton** [comment]
Thanks, I’ll check this out soon and make a docs patch once I get it working.

**Comment by danielcompton** [comment]
I forgot to mention that clj-kondo is running as part of a lint script which which runs a few different linters and pipes them all to [reviewdog](https://github.com/reviewdog/reviewdog). I had an attempt at this, and your caching strategy seems like it would work in CI. However it won't invalidate the cache when running the script locally, unless I also reimplement CircleCI's caching logic. 

I'm working on local cache invalidation in the script, but it seems like it would be useful if this was part of clj-kondo?

**Comment by danielcompton** [comment]
Here's the script I came up with. It works, but I think it will be a little bit brittle:

```bash
if [[ "$CI" = "true" ]]; then
    # Provide the full path to allow reviewdog to localise it.
    # Only include files that don't start with a fully qualified path /
    clj_kondo_config='{:output {:canonical-paths true :include-files ["^[^\\/]"]}}'
else
    clj_kondo_config='{:output {:include-files ["^[^\\/]"]}}'
fi

# If our deps.edn file hasn't changed, then reuse the cache
if cmp -s <(shasum deps.edn) .clj-kondo/.cache/checksum; then
	git ls-files | grep -E '.clj[cs]?$' | xargs clj-kondo --config "${clj_kondo_config}" --lint
else
	# Otherwise, lint everything on the classpath (this provides extra info when linting our source)
	clj-kondo --lint "$(clojure -Spath -A:dev:test)" --config "${clj_kondo_config}"
	# Checksum our dependencies so we don't rebuild this cache this unnecessarily
	shasum deps.edn > .clj-kondo/.cache/checksum
fi
```

**Comment by borkdude** [comment]
> It seems like it would be useful if this was part of clj-kondo

Right now the cache logic is straightforward. The cache is only used to save things like arities so when you edit namespace `x` but are using namespace `y` from it, we can consult data from namespace `y` which we've seen before. Cache is maybe not the right name, but database maybe also isn't. Naming is hard.

I'd like to keep the caching logic as simple as possible for as long as possible. As you may know, next to naming, cache invalidation is one of the hardest problem in CS.

> Here's the script I came up with. It works, but I think it will be a little bit brittle

Why do you think this is brittle? Looks good to me?

**Comment by borkdude** [comment]
One additional note about caching: it would be good to also include the clj-kondo version (produced by `clj-kondo --version`) in the checksum as clj-kondo uses that also in the `.cache` directory structure.

One additional note about speed: you could try linting with the uberjar instead of the executable in CI if things are taking longer than say, 10 seconds. The java version is faster overall, but has worse startup than the executable.

**Comment by danielcompton** [comment]
> One additional note about caching: it would be good to also include the clj-kondo version (produced by clj-kondo --version) in the checksum as clj-kondo uses that also in the .cache directory structure.

Yeah I just realised this too, and I'll also need to include the clj-kondo version in the CircleCI checksum. This is what I meant by brittle, I need to encode a bunch of knowledge about clj-kondo into a shell script wrapper around it.

**Comment by borkdude** [comment]
It's not perfect, but I think it's not that bad either. For now I prefer this over changing the logic inside clj-kondo as I'm concerned that will lead to hard-to-debug problems.

**Comment by borkdude** [comment]
Another idea: 

If you use clj-kondo as a git or maven dependency in deps.edn, you don't have to bother with the version, since that's already part of the deps.edn file then. You can then invoke it with `clojure -A:clj-kondo --lint ...`.

**Comment by danielcompton** [comment]
That's a good idea, I'l that some more. We like to use the native-image version of clj-kondo as that's faster for quick one-off builds, and the slightly slower execution in CI isn't a big deal, especially once we cache analysis.

I completely understand wanting to keep the core simple, I'll keep noodling on caching and cache invalidation.

**Comment by borkdude** [comment]
@danielcompton I translated your bash script to Clojure. Using [babashka](https://github.com/borkdude/babashka/) it can be run fast in CI.

``` clojure
(require
 '[clojure.java.shell :refer [sh]]
 '[clojure.java.io :as io]
 '[clojure.string :as str])

(def clj-kondo-config
  (if (System/getenv "CI")
    ;; Provide the full path to allow reviewdog to localise it.
    ;; Only include files that don't start with a fully qualified path /
    {:output {:canonical-paths true :include-files ["^[^\\/]"]}}
    {:output {:include-files ["^[^\\/]"]}}))

(def checksum-file (io/file ".clj-kondo" ".cache" "checksum"))

(defn lint-sources []
  (println "Linting only sources.")
  (as-> (sh "git" "ls-files") $
    (:out $)
    (sh "grep" "-E" ".clj[cs]?$" :in $)
    (str/split (:out $) #"\n")
    (apply sh "clj-kondo" "--config" (str clj-kondo-config) "--lint" $)
    (println (:out $))))

(defn checksum []
  (let [clj-kondo-version (:out (sh "clj-kondo" "--version"))
        clj-kondo-checksum (:out (sh "shasum" :in clj-kondo-version))
        deps-edn-checksum (:out (sh "shasum" "deps.edn"))]
    (str/join [clj-kondo-checksum deps-edn-checksum])))

(defn write-checksum []
  (spit checksum-file (checksum)))

(defn lint-all []
  (println "Linting deps + sources.")
  (let [classpath (:out (sh "clojure" "-Spath" "-A:dev:test"))
        results (sh "clj-kondo" "--lint" classpath "--config" (str clj-kondo-config))]
    (println (:out results)))
  (write-checksum))

(if (and (.exists checksum-file)
         (= (checksum) (slurp checksum-file)))
  (lint-sources)
  (lint-all))
```

Note that you can get the paths from a `deps.edn` file relatively easily:

``` clojure
(def deps-edn (edn/read-string (slurp "deps.edn")))
(def paths (into (:paths deps-edn)
                 (-> deps-edn :aliases :test :extra-paths)))
```

**Comment by borkdude** [comment]
A snippet to extract paths from `deps.edn`:

```
(def deps-edn (edn/read-string (slurp "deps.edn")))
(def paths (into (:paths deps-edn)
                 (-> deps-edn :aliases :test :extra-paths)))
```

**Comment by SevereOverfl0w** [comment]
The cache could be checksummed off the .cpcache/*.basis file that is created by clj.

**Comment by borkdude** [comment]
Another idea: clj-kondo could keep a global cache for .jars which can be re-used among projects.

**Comment by borkdude** [comment]
@danielcompton @SevereOverfl0w For now I have implemented the following on master:

When invoking clj-kondo with `--no-warnings` (which indicates you are linting only for populating the cache), `.jar` files will be skipped if they have a non-SNAPSHOT version:

```
$ clojure -M:clj-kondo --no-warnings --lint $(clojure -Spath)
$ clojure -M:clj-kondo --no-warnings --lint $(clojure -Spath)
$ clojure -M:clj-kondo --no-warnings --lint $(clojure -Spath)
sci-0.1.0.jar was already linted, skipping
cheshire-5.10.0.jar was already linted, skipping
transit-clj-1.0.324.jar was already linted, skipping
datalog-parser-0.1.7.jar was already linted, skipping
bencode-1.1.0.jar was already linted, skipping
clojure-1.10.1.jar was already linted, skipping
edamame-0.0.11-alpha.12.jar was already linted, skipping
sci.impl.reflector-0.0.1.jar was already linted, skipping
tools.reader-1.3.2.jar was already linted, skipping
jackson-core-2.10.2.jar was already linted, skipping
jackson-dataformat-cbor-2.10.2.jar was already linted, skipping
jackson-dataformat-smile-2.10.2.jar was already linted, skipping
tigris-0.1.2.jar was already linted, skipping
transit-java-1.0.343.jar was already linted, skipping
core.specs.alpha-0.2.44.jar was already linted, skipping
spec.alpha-0.2.176.jar was already linted, skipping
commons-codec-1.10.jar was already linted, skipping
jaxb-api-2.3.0.jar was already linted, skipping
msgpack-0.6.12.jar was already linted, skipping
json-simple-1.1.1.jar was already linted, skipping
javassist-3.18.1-GA.jar was already linted, skipping
```

We could extend this to gitlibs as well.

**Comment by borkdude** [comment]
One problem with the above is that the jar name has to be unique, but we can't reliably discover the root of the .m2 repo (to fabricate a unique jar name) since org names can have arbitrary depth.

**Comment by borkdude** [comment]
A solution to the above: clj-kondo writes the full path of the linted jar to the skip entry in the cache and compares it the next time. If it's different, the jar gets re-linted, if it's the same, skipped.

```
$ cat .clj-kondo/.cache/2020.10.11-SNAPSHOT/skip/edamame-0.0.11-alpha.12.jar
/Users/borkdude/.m2/repository/borkdude/edamame/0.0.11-alpha.12/edamame-0.0.11-alpha.12.jar%
```
